### PR TITLE
Re Issue #3962: Make auto templates for a Taxonomy Pod behave sensibly

### PR DESCRIPTION
**Pull request by gwhitney**
I am submitting this pull request as suggested in the discussion to Issue #3962, where I requested that auto templates for taxonomy pods not be turned off. Let me start by saying that although I have considerable coding experience, I am relatively inexperienced with git workflows, so my apologies if I do not have the forking/branching structure correct, and if so, I am happy to close this pull request and create  a new better one, with some guidance as to what should be changed.

Next, I should say that this pull request is organized as a series of successive recommendations regarding auto template handling for taxonomy types; you may decide to take almost any initial sequence of these commits and leave the remainder for further deliberation or reject the remainder. 

Finally, let me explain the contents/motivation of each commit further, so you can better decide which to take, despite the risk of tl;dr. (Of course, I personally use and recommend all of these changes.) This list is in the order of the commits in the request:

59d2e7e -- this is just the suggested patch from the discussion concerning #3962. Just by itself it makes auto templates with taxonomies more usable and understandable, in that it changes the default type of the pod created to run the template against from the pod type of the _query_ that generated the post(s) to the post type of the _post_ that is actually currently being displayed.

f201889 -- Another simple change that extends the usefulness of auto templates with taxonomies to those taxonomies which span multiple post types.. By including a magic code like `{@post_type}` in the auto template name, one can use different templates for different post types that might show up in the taxonomy archive.
 
90a5f88 -- I noticed that 'the_content' was hard-coded into the temporary hook removal and re-installation in `load_template`, even though the actual filter used may have been set to something else, and that the priority of the hook was changing from 10.5 to 10. So instead, store the filters which have been hooked and their priorities in a member variable so that they can be maintained through `load_template`. Does not provide any new functionality, but is necessary for following commits, since they are just slightly more intensive in terms of setting/removing hooks.

4b78da6 -- auto templates were not firing in search results, leading to an inconsistent look for post summaries between taxonomy archives and search results. The issue was that no post_type can be inferred from the wp_query when it is a search query -- it is truly difficult to tell what types of posts might be turned up in the search. Therefore, this commit implements a mechanism by which if `$current_post_type` is undetermined by the wp_query outside the loop, it can be recomputed for each post inside the loop by making use of the post_type of the current post. Note that this change does _not_ alter the `$current_post_type` if it was determined by the wp_query itself, e.g. in case of a taxonomy archive; that behavior preserves the operation of auto templating for taxonomy archives.

The remaining five commits really go together as more of a unit, even though they could be teased apart (and the commit messages do give some more detail about what's in each). The underlying idea of these five commits is to allow auto templates for taxonomies also to affect the taxonomy description that shows up at the top of the taxonomy archive (in most/many themes), since that's what I assumed it _would_ modify when I first started using Pods, and it is in fact handy to be able to incorporate custom fields into that description. The mechanism for doing so consists of: allowing the "single" templates for taxonomies as well; arranging that the "single" template for a taxonomy filters `get_the_archive_description` by default (note that filtering `get_the_archive_title` is another totally reasonable use which now can be specified in the auto template settings tab); allowing both the "single" and "archive" filters to be in place simultaneously so that the single (for the description) can apply outside the loop and the archive one can apply inside the loop; arranging that when a taxonomy "single" template runs, it runs against the item (term) in that Taxonomy Pod corresponding to the term of the query; and allowing there to be no "single" template, for the case in which the archive posts should be templated but the description doesn't need to be.

I understand that as a result, there is a lot going on in this pull request. I hope that doesn't reduce the chances of at least the first commit being adopted, which already has significant benefit. And I hope that you are able to consider the full package; all of the changes are in the spirit of that first one, and I believe overall they make the auto template feature significantly more powerful while also making it simpler and more understandable: single templates affect the presentation of the Pod items themselves, archive templates affect the post archives associated with the Pod; for Custom Post Type Pods these are the post archives, as well as search results, and for Taxonomy Pods, these are the Taxonomy archives. Thus, a given post of a Custom Post Type Pod can have appearance A in a single post page, appearance B in the post archive of that post type and in search results, and appearance C in the taxonomy archive of one associated Taxonomy Pod, appearance D in another Taxonomy, and so on. 

In case it's helpful for understanding the changes and their motivation, I use them for content of a publication. The custom post types are different sorts of content (article, letter, etc.). There are taxonomies for the contributor of the content, what issue it is in, what topics it relates to, etc. Articles are presented differently in the list of all articles than in the page for a given contributor (e.g., there's no need to mention the contributor in the latter case). And all of the content types look very different in the page for a given issue, which is (a mockup of) the physical layout of the paper version of the publication.

Hope this is of benefit and interest to you and the community of Pods users.

### CODE REVIEW

**Comment by gwhitney** [comment]
Oh dear, I wasn't aware of pull request #4696 when I created this one. This will certainly not merge effortlessly with #4696. It does however address the identical issue, specifically in the final commit of this pull request. The separation of the code determining the name of the filter to use is orthogonal to this pull request, and is potentially a worthwhile cleanup on its own. So there are various possibilities, with the leading candidate as far as I can see is for me to close and re-do this pull request once #4696 has been merged (although as I say this could be taken instead of #4696 as it addresses the same issue). I will leave this pull request open for now, so that you can take a look and provide feedback on the direction to take with it. If indeed the desired route is for me to close and re-do this after #4696 is merged, please let me know which portions of the changes in this pull request are desired to be included in the re-do. Thanks!

**Comment by jimtrue** [comment]
@gwhitney I got @sc0ttkclark to look at the conflicts from #4696 as soon as the Travis checks are finished on this one, I'll try to pull down your fix and test it locally.

**Comment by sc0ttkclark** [comment]
FYI I only resolved the conflicts that came out of the phpcs formatting merge. I still need to look at the conflicts between this and that other PR

**Comment by sc0ttkclark** [comment]
This looks good, needs some testing to confirm it works as expected / intended.

File: components/Templates/includes/auto-template/Pods_Templates_Auto_Template_Front_End.php [hunk 1]
```diff
@@ -47,7 +56,7 @@ public function get_pod_filter( $current_post_type = '', $possible_pods = array(
 		if ( isset( $possible_pods[ $current_post_type ] ) ) {
 			$this_pod = $possible_pods[ $current_post_type ];
 
-			if ( is_singular( $current_post_type ) ) {
+      if ( is_singular( $current_post_type ) ) {
```
Comment (GaryJones) [inline components/Templates/includes/auto-template/Pods_Templates_Auto_Template_Front_End.php]: This spaces indentation is going against code standards.

Comment (sc0ttkclark) [reply components/Templates/includes/auto-template/Pods_Templates_Auto_Template_Front_End.php]: Fixed

**Comment by gwhitney** [comment]
Wow, thanks for the activity on this. Let me know if I can be of assistance in this effort.

**Comment by jimtrue** [comment]
@gwhitney I'm setting up my local for testing this in all the ways you discussed above. I'll reach out to you if I need assistance. Are you on our Pods Slack Chat at https://pods.io/chat/ ?

**Comment by gwhitney** [comment]
Well, I am teaching at a university this semester, so I have not been keeping too up-to-date with my various programming projects. I don't think I am on the Slack chat; it's certainly not someplace I go visit, and I don't think I am getting any kind of email notices from a Slack chat (I only have a fairly vague idea what Slack is...)

**Comment by jimtrue** [comment]
No worries. I'll try to focus my conversation with you here and give succinct, specific questions. Thanks for all your time on this!

**Comment by jimtrue** [comment]
@gwhitney Can you re-merge your branch with our updated 2.7.10 (or 2.x); I was getting ready to test this one last week and we had another item that was conflicting. Since I need to pull down your entire fork to test, I need it merged again with our latest code.

Thanks!

**Comment by sc0ttkclark** [comment]
Merging this into a temporary holding branch under our control for easier handling going forward

**Pull request merged**